Closes #298 - Metrics are now consistent with Go

jpkrohling opened the pull request:
Normalizes the metrics on this library with the ones from the Go library, not only in the types of metrics and tags, but also in the names (underscores instead of dashes, for instance). One metric on the Java library was hurt in the process and didn't survive.

Includes also the commits from @NeoCN , with the initial Micrometer integration.

jpkrohling commented:
Travis is currently failing due to an unrelated issue. I just re-run a build for #326 which was all green and it failed on the same step as this PR here is failing.

eundoosong commented on jaeger-core/src/main/java/com/uber/jaeger/metrics/Metrics.java:
```diff
@@ -32,7 +32,7 @@ public Metrics(StatsFactory factory) {
       }
 
-      StringBuilder metricBuilder = new StringBuilder("jaeger.");
+      StringBuilder metricBuilder = new StringBuilder("jaeger:");
```
I have a question, 
go-client starts with "metric:",  don't we need to be consistent with go-client?

jpkrohling replied:
Yes, the same metrics reported via go-client should have the same name as the java-client. I actually compared the metric names + the output from here: https://github.com/jaegertracing/jaeger/issues/572 . Do you mean that metrics aren't shown like this:

```
# HELP jaeger:finished_spans jaeger:finished_spans
# TYPE jaeger:finished_spans counter
jaeger:finished_spans 0
# HELP jaeger:reporter_queue_length jaeger:reporter_queue_length
# TYPE jaeger:reporter_queue_length gauge
jaeger:reporter_queue_length 0
# HELP jaeger:reporter_spans jaeger:reporter_spans
# TYPE jaeger:reporter_spans counter
jaeger:reporter_spans{result="dropped"} 0
jaeger:reporter_spans{result="err"} 0
jaeger:reporter_spans{result="ok"} 0
```

eundoosong replied:
I think you're right. 
From go-client,  "metric:" is used for extracting tags from struct Metrics.
"jaeger" is used for metric name. :)

checketts replied:
I realize this PR was to unify with the Go client. However by departing from the dot notation with micrometer you lose the hierarchical nature for config.  That way users can `management.metrics.enable.jaegar=false` if they want to disable/customize those metrics.

I also recommend using dots in the metric names like `@Metric(name = "finished.spans")`  The end result to Prometheus will replace the dots with underscores (or format the metric to match the backend).

yurishkuro replied:
> That way users can management.metrics.enable.jaegar=false if they want to disable/customize those metrics.

@checketts could you please elaborate on this? Where would you use this ^ dot-delimited syntax?

checketts replied:
For Spring Boot applications you can set those property to control which metrics you want to enable/disable. Similar to enabling/disabling logging levels.

For non-Spring applications then the user would add in their own MeterFilter like so `myRegistry.config().filter(MeterFilter.denyNameStartsWith("jaegar.reporter"))`. In that case the colon wouldn't break the usage, but it does deviate from the expected notation.

yurishkuro replied:
mentioned this in https://github.com/jaegertracing/jaeger/issues/572#issuecomment-372483488

what if everything was separated with underscore?

checketts replied:
Only dots imply the hierarchical naming for Micrometer config. If using a colon or underscore, then there is no tokenization and each metric name is treated as a single value.  

So I believe the Spring property would be: `management.metrics.enable.jaeger:reporter_spans: false` but yaml will complain with colon's in the key so it would need to be escaped.

If the colon was replaced then it will avoid the yaml problems. But it will still not be tokenized so Spring Users would have to disable each metric individually.

Just to be clear: A metric named `jaeger.finished.spans` in Micrometer will be output as `jaeger_finished_spans` in Prometheus. So it shouldn't impact that aspect of #572

yurishkuro replied:
@checketts what is the purpose of enabling/disabling metrics via config file?

yurishkuro replied:
Also you said 

> I also recommend using dots in the metric names like `@Metric(name = "finished.spans")`

Finished spans is a single metric, why do you gain by treating that name as hierarchical?

fwiw I am trying to reconcile conflicting asks (see https://github.com/jaegertracing/jaeger/issues/732)

checketts replied:
The purpose of enabling/disabling metrics via config file is similar to enabling/disabling debug logging. The user of client can disable metrics they don't need (or want to pay for when it comes to 3rd party monitoring systems like DataDog or CloudWatch).  

The dot name separation supports hierarchy, since to Micrometer's view the metric's complete name is `jaeger.finished.spans`, so a user would be able to enable/disable all the Jaeger metrics easily. Specifically regarding the `finished.spans` section, Micrometer formats the metric name according to the corresponding backend (underscores for Prometheus, camel case for Atlas, etc.)  

I don't think the dot notion will affect #732 much since Micrometer writes out the metrics, trying to match the corresponding backend. So the underscores will be preserved on output to Prometheus.

The colon is really the only sticking point. For users than really want the colon, they could add a `MeterFilter` like so:

```
registry.config().filter(new MeterFilter() {
  public Meter.Id map(Meter.Id id) {
    if(id.getName().startsWith("jaegar.")) {
      return new Meter.Id(id.getName().replace("jaegar.", "jaegar:"), allTags, id.getBaseUnit(), id.getDescription(), id.getType());
    }
    
    return id;
  }
});
```

You could even include that `MeterFilter` in your client and let the user wire it up if they need it.  

(As a  I'm also not a fan of the colon since it is typically used in [recording rules](https://prometheus.io/docs/practices/rules/#examples) in Prometheus and implies an aggregation has been applied)

eundoosong commented on jaeger-core/src/main/java/com/uber/jaeger/metrics/Metrics.java:
```diff
@@ -93,147 +95,110 @@ public static Metrics fromStatsReporter(StatsReporter reporter) {
+  // Number of times the Sampler succeeded to retrieve and update sampling strategy
+  public Counter samplerUpdated;
+
+  @Metric(name = "sampler_updates", tags = @Tag(key = "result", value = "ok"))
```
I think value should be "err".

jpkrohling replied:
Nice catch!

objectiser commented on jaeger-micrometer/src/test/java/com/uber/jaeger/micrometer/MicrometerTest.java:
```diff
@@ -0,0 +1,75 @@
+    metrics.decodingErrors.inc(1);
+    assertThat(metricsReporter.registry()
+          .get("jaeger:span_context_decoding_errors").counter().count(),
+          IsEqual.equalTo(1D)
```
Curious why incrementing the decoding errors by 1 translates to 10 in the counter? Same applies in following tests.
If there is a good reason, might be worth documenting it in the test, as doesn't look right.

yurishkuro commented on jaeger-micrometer/src/main/java/com/uber/jaeger/micrometer/MicrometerStatsReporter.java:
```diff
@@ -0,0 +1,67 @@
+import lombok.extern.slf4j.Slf4j;
+
+@Slf4j
+public class MicrometerStatsReporter implements StatsReporter {
```
StatsReporter should be deprecated, because it's very inefficient. StatsFactory is the preferred method (which we may also want to deprecate and rename to MetricsFactory to be consistent with Go)

jpkrohling replied:
#338 created for this

yurishkuro replied:
Sorry, my point (besides deprecating the StatsXXX API) was that this class should implement StatsFactory, not StatsReporter. The main Metrics class uses the factory to create stats objects once and call methods on them. In contrast, if this implements the StatsReporter you have to do very expensive `translateTags(tags)` for every `incCounter`, making it orders of magnitude more expensive call.

jpkrohling replied:
I see what you mean now. I'll deprecate the StatsReporter as part of #338, but I changed this commit to get it to implement `StatsFactory`

jpkrohling commented:
@yurishkuro , @objectiser : I just updated the last commit to include a small change to the README, showing how to use the stats reporter. I also fixed the issue @eundoosong spotted. About deprecating the StatsReporter, I'd leave it outside of this PR. I created #338 for that.

objectiser commented on jaeger-core/README.md:
```diff
@@ -76,6 +74,17 @@ environment approach described above.
 
 More information about using the `TracerResolver` can be found [here](../jaeger-tracerresolver/README.md).
 
+#### Reporting internal stats via Micrometer
```
nit: stats -> metrics

jpkrohling commented:
@yurishkuro, @objectiser, this is now ready for a final review/merge. Includes the change to use `StatsFactory` instead of `StatsReporter`.

objectiser commented on jaeger-core/README.md:
````diff
@@ -76,6 +74,17 @@ environment approach described above.
+`MicrometerStatsReporter` like this:
+
+```java
+MicrometerStatsReporter metricsReporter = new MicrometerStatsReporter();
````
This example needs to be updated to use the `MicrometerStatsFactory` approach.

jpkrohling replied:
Fixed

pavolloffay commented on jaeger-core/src/main/java/com/uber/jaeger/metrics/Metrics.java:
```diff
@@ -80,10 +80,12 @@ public static String addTagsToMetricName(String name, Map<String, String> tags)
-      sb.append(key);
-      sb.append("=");
-      sb.append(tags.get(key));
+      if (!tags.get(key).isEmpty()) {
```
nit: there are two accesses to the map. `SortedMap` can be used Instead of `SortedSet` and then just iterate over entries.

jpkrohling replied:
I don't think this belongs to this PR. Could you open an issue to change this?

pavolloffay replied:
I will directly open a PR. 

Why does it have to be sorted?

jpkrohling commented:
@yurishkuro , is there something that needs to be addressed to get this PR merged?

yurishkuro commented on jaeger-micrometer/src/main/java/com/uber/jaeger/micrometer/MicrometerStatsFactory.java:
```diff
@@ -0,0 +1,83 @@
+
+      @Override
+      public void inc(long delta) {
+        log.trace("Incrementing counter {} by {}", name, delta);
```
please remove logging on metrics

jpkrohling replied:
I can remove it, but this call is a noop after a few cycles, as hotspot-based JVMs (and probably others) will just use branch prediction to skip it once it sees that it's never executed.

Or are you concerned about something else?

jpkrohling commented:
Is this ready to be merged, or is there anything else missing?

yurishkuro commented on jaeger-core/README.md:
```diff
@@ -76,6 +74,20 @@ environment approach described above.
+
+The Jaeger Java Client collects internal metrics and is able to report them via [Micrometer](http://micrometer.io).
+To accomplish that, include the artifact `com.uber.jaeger:jaeger-micrometer` as a dependency to your project and use
+`MicrometerStatsReporter` like this:
```
typo: s/MicrometerStatsReporter/MicrometerStatsFactory/

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/metrics/Metrics.java:
```diff
@@ -95,147 +95,110 @@ public static Metrics fromStatsReporter(StatsReporter reporter) {
-  )
+  @Metric(name = "started_spans", tags = @Tag(key = "sampled", value = "y"))
   // Number of sampled spans started by this tracer
   public Counter spansStarted;
```
can we rename this to spansStartedSampled?

jpkrohling replied:
Done

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/metrics/Metrics.java:
```diff
@@ -95,147 +95,110 @@ public static Metrics fromStatsReporter(StatsReporter reporter) {
-  // Number of not-sampled spans started by this tracer
+  @Metric(name = "started_spans", tags = @Tag(key = "sampled", value = "n"))
+  // Number of unsampled spans started by this tracer
   public Counter spansNotSampled;
```
can we rename this to spansStartedNotSampled?

jpkrohling replied:
Done

yurishkuro commented on jaeger-core/src/main/java/com/uber/jaeger/Tracer.java:
```diff
@@ -407,12 +407,10 @@ private String debugId() {
+        metrics.spansStartedNotSampled.inc(1);
       }
-      metrics.spansStarted.inc(1);
+      metrics.spansStartedSampled.inc(1);
```
this should be on the `else` part

yurishkuro replied:
is the test wrong?

jpkrohling replied:
Good catch. There was no test for this case, so, I fixed this and added a test. Originally, the code had this:

```
      if (context.isSampled()) {
        metrics.spansSampled.inc(1);
      } else {
        metrics.spansNotSampled.inc(1);
      }
metrics.spansStarted.inc(1);
```

A previous iteration removed the `spansSampled` and kept the `spansStarted`, but looks like the correct is to have `spansSampled` and `spansNotSampled`, removing `spansStarted`.
